File: pt-4/247.strobogrammatic-number-ii.go
package leetcode

// FindStrobogrammatic247 builds numbers from the outside inward using valid
// rotated digit pairs. The outermost layer skips 0 unless the total length is 1.
//
// Time/space: proportional to output size.
func FindStrobogrammatic247(n int) []string {
	pairs := [][2]string{{"0", "0"}, {"1", "1"}, {"6", "9"}, {"8", "8"}, {"9", "6"}}
	var build func(int, int) []string
	build = func(length, total int) []string {
		if length == 0 {
			return []string{""}
		}
		if length == 1 {
			return []string{"0", "1", "8"}
		}
		ans := []string{}
		for _, inner := range build(length-2, total) {
			for _, p := range pairs {
				if length == total && p[0] == "0" {
					continue
				}
				ans = append(ans, p[0]+inner+p[1])
			}
		}
		return ans
	}
	return build(n, n)
}

File: companies/Uber/247.strobogrammatic-number-ii.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<string> findStrobogrammatic(int n) {
        /*
        Approach: recursively build from the inside out. For each inner string,
        wrap it with valid rotated digit pairs. Skip leading zero pairs at the
        outermost level.

        Complexity: O(5^(n/2) * n) time and output space.
        */
        vector<pair<char,char>> pairs = {{'0','0'}, {'1','1'}, {'6','9'}, {'8','8'}, {'9','6'}};
        function<vector<string>(int, int)> build = [&](int length, int total) -> vector<string> {
            if (length == 0) return {""};
            if (length == 1) return {"0", "1", "8"};
            vector<string> ans;
            for (const string& inner : build(length - 2, total)) {
                for (auto [a, b] : pairs) {
                    if (length == total && a == '0') continue;
                    ans.push_back(string(1, a) + inner + string(1, b));
                }
            }
            return ans;
        };
        return build(n, n);
    }
};
